### PR TITLE
udev-node: remove stack directories and their lock files by workers, rather than manager

### LINKED ISSUE #34637: udev/pid1: loopback diskseq symlinks in /run/ are created for each run and kept around until a restart cycle ends leading to inode exhaustion
If a unit uses a loop device for any reason, a diskseq symlink is created in /run/udev/links/ for each invocation. If a unit fails and it has Restart= set, the symlinks are all kept around until a rate limit is hit, or the unit is manually stopped, and only then they are cleared. If a unit sets StartlimitIntervalSec=0 to disable rate limiting, and always fail, eventually /run/ runs out of inodes due to the symlinks.

To reproduce:

```
mkdir /var/foo/bar -p
mksquashfs /var/foo /var/foo.raw
cat <<EOF> /etc/systemd/system/test.service
[Unit]
StartLimitIntervalSec=0
[Service]
ExecStart=false
Restart=on-failure
MountImages=/var/foo.raw:/var/foo
EOF
```

Then start the test.service and observe the number of links grow indefinitely `ls /run/udev/links | grep diskseq | wc -l`

While this happens, udevadm monitor shows a stream of:

```
UDEV  [849.134231] change   /devices/virtual/block/loop4 (block)
UDEV  [849.187554] change   /devices/virtual/block/loop4 (block)
KERNEL[849.370486] change   /devices/virtual/block/loop3 (block)
UDEV  [849.425013] change   /devices/virtual/block/loop3 (block)
KERNEL[849.448361] change   /devices/virtual/block/loop4 (block)
UDEV  [849.486913] change   /devices/virtual/block/loop4 (block)
```

udevadm info on the loop device:

```
P: /devices/virtual/block/loop0
M: loop0
R: 0
J: b7:0
U: block
T: disk
D: b 7:0
N: loop0
L: 0
S: disk/by-diskseq/478
Q: 479
E: DEVPATH=/devices/virtual/block/loop0
E: DEVNAME=/dev/loop0
E: DEVTYPE=disk
E: DISKSEQ=479
E: MAJOR=7
E: MINOR=0
E: SUBSYSTEM=block
E: USEC_INITIALIZED=5176916
E: NVME_HOST_IFACE=none
E: SYSTEMD_READY=0
E: DEVLINKS=/dev/disk/by-diskseq/478
E: TAGS=:systemd:
E: CURRENT_TAGS=:systemd:
```

Stepping through with gdb, I pin pointed the exact moment when the new link is created, and it's when the loop_device_unref from here is called:

https://github.com/systemd/systemd/blob/598bb6fde46550ce020dd1d9853b316b766cb923/src/shared/dissect-image.c#L4040

more precisely when `d->fd` is closed here:

https://github.com/systemd/systemd/blob/598bb6fde46550ce020dd1d9853b316b766cb923/src/shared/loop-util.c#L812

It feels like we are keeping <some> reference somewhere that prevents udev from cleaning up these links while a restart cycle is ongoing

### PR DESCRIPTION
**Pull request by yuwata**
This effectively reverts #34644.
Replaces #34664.
Closes #34664.
Fixes #34637.

### CODE REVIEW

File: src/udev/udev-node.c [hunk 1]
```diff
@@ -514,8 +518,10 @@ static int link_update_diskseq(sd_device *dev, const char *slink, bool add) {
 }
 
 static int link_update(sd_device *dev, const char *slink, bool add) {
+        _cleanup_(rmdir_and_freep) char *path = NULL; /* #1 */
```
Comment (poettering) [inline src/udev/udev-node.c]: "path" might be a bit generic, could be the dir or the lock file, maybe be explicit here, i.e. dirpath? or so?

**Comment by poettering** [comment]
This is sooo much nicer! Thanks!

**Comment by yuwata** [comment]
> Does this mean the links are still created, but then deleted later?

RIght.

**Comment by bluca** [comment]
Sorry but I don't think that's safe - due to the spam of links causing resource exhaustion on /run/ it might not even be possible to start workers or anything else. We should not do work that is unneeded, it is wasteful and risky, and causing systems to lock up and need a hard reboot is really bad publicity. So please drop the second commit. The other two look like a great improvement on their own, no issue with them.

**Comment by poettering** [comment]
> Sorry but I don't think that's safe - due to the spam of links causing resource exhaustion on /run/ it might not even be possible to start workers or anything else. We should not do work that is unneeded, it is wasteful and risky, and causing systems to lock up and need a hard reboot is really bad publicity. So please drop the second commit. The other two look like a great improvement on their own, no issue with them.

Uh? if we clean up symlinks immediately (as with this PR), then these dirs will be always empty. The special casing of the diskseq stuff is not necessary.

**Comment by yuwata** [comment]
> Sorry but I don't think that's safe - due to the spam of links causing resource exhaustion on /run/ it might not even be possible to start workers or anything else. We should not do work that is unneeded, it is wasteful and risky, and causing systems to lock up and need a hard reboot is really bad publicity. So please drop the second commit. The other two look like a great improvement on their own, no issue with them.

Now, for by-diskseq, symlink under /run/udev/links/ is created at most one per device at the same time (yes, zero with the previous fix). See the change for the test case. To consume whole inodes of /run/, it needs huge number of disks and partitions, and in such situation, something broken before /run/ becomes full, e.g. #34690.

**Comment by yuwata** [comment]
@bluca Anyway, please try to torture udev with this patch. At least, the reproducer of the original issue cannot fill /run/.

**Comment by bluca** [comment]
> > Sorry but I don't think that's safe - due to the spam of links causing resource exhaustion on /run/ it might not even be possible to start workers or anything else. We should not do work that is unneeded, it is wasteful and risky, and causing systems to lock up and need a hard reboot is really bad publicity. So please drop the second commit. The other two look like a great improvement on their own, no issue with them.
> 
> Now, for by-diskseq, symlink under /run/udev/links/ is created at most one per device at the same time (yes, zero with the previous fix). See the change for the test case. To consume whole inodes of /run/, it needs huge number of disks and partitions, and in such situation, something broken before /run/ becomes full, e.g. #34690.

But that's till async, no? So in fact there's no guarantee of that, and it relies on the workers being able to be spawned and run, right?

**Comment by poettering** [comment]
> But that's till async, no? So in fact there's no guarantee of that, and it relies on the workers being able to be spawned and run, right?

No. It's done after the symlinks are removed/added in the workers, i.e. right where it should be.

**Comment by bluca** [comment]
CI seems unhappy though, or is this a pre-existing condition? I don't recall seeing it before:

```
[  285.398842] TEST-64-UDEV-STORAGE.sh[682]: + [[ /dev/sdf8 != \/\d\e\v\/\s\d\e\8 ]]
[  285.399188] TEST-64-UDEV-STORAGE.sh[682]: + echo 'ERROR: symlink '\''/dev/disk/by-partlabel/test8'\'' points to '\''/dev/sdf8'\'' but '\''/dev/sde8'\'' was expected'
[  285.399619] TEST-64-UDEV-STORAGE.sh[682]: ERROR: symlink '/dev/disk/by-partlabel/test8' points to '/dev/sdf8' but '/dev/sde8' was expected
```

**Comment by yuwata** [comment]
> CI seems unhappy though, or is this a pre-existing condition? I don't recall seeing it before:
> 
> ```
> [  285.398842] TEST-64-UDEV-STORAGE.sh[682]: + [[ /dev/sdf8 != \/\d\e\v\/\s\d\e\8 ]]
> [  285.399188] TEST-64-UDEV-STORAGE.sh[682]: + echo 'ERROR: symlink '\''/dev/disk/by-partlabel/test8'\'' points to '\''/dev/sdf8'\'' but '\''/dev/sde8'\'' was expected'
> [  285.399619] TEST-64-UDEV-STORAGE.sh[682]: ERROR: symlink '/dev/disk/by-partlabel/test8' points to '/dev/sdf8' but '/dev/sde8' was expected
> ```

Ouch, I will take a look...

**Comment by yuwata** [comment]
Updated.
Now, lock file is moved to /run/udev/.links.lock/, as workers need to take a lock before creating the directory under /run/udev/links, and release lock after trying to remove the directory.
PTAL.

**Comment by yuwata** [comment]
Arch mkosi failure should be unrelated, and already reported at #34073.

<details>
<summary>logs</summary>

```
 6/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-multipath_basic_failover                 FAIL             30.49s   exit status 1
>>> MALLOC_PERTURB_=198 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 SYSTEMD_INTEGRATION_TESTS=1 /home/runner/work/systemd/systemd/test/integration-test-wrapper.py --meson-source-dir /home/runner/work/systemd/systemd --meson-build-dir /mnt/mkosi/build --name TEST-64-UDEV-STORAGE-multipath_basic_failover --storage volatile --firmware linux --exit-code 123 --unit TEST-64-UDEV-STORAGE.service --vm --mkosi /usr/bin/mkosi -- --kernel-command-line-extra=systemd.setenv=TEST_FUNCTION_NAME=testcase_multipath_basic_failover --configure-script /mnt/mkosi/build/../../../home/runner/work/systemd/systemd/test/TEST-64-UDEV-STORAGE/multipath_basic_failover.configure
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Listing only the last 100 lines from a long log.
[   21.078344] TEST-64-UDEV-STORAGE.sh[726]: + echo TEST_FUNCTION_NAME=testcase_multipath_basic_failover
[   21.080965] TEST-64-UDEV-STORAGE.sh[726]: TEST_FUNCTION_NAME=testcase_multipath_basic_failover
[   21.081171] TEST-64-UDEV-STORAGE.sh[726]: + testcase_multipath_basic_failover
[   21.081638] TEST-64-UDEV-STORAGE.sh[726]: + local dmpath i path wwid
[   21.081867] TEST-64-UDEV-STORAGE.sh[726]: + cat
[   21.085864] TEST-64-UDEV-STORAGE.sh[726]: + udevadm lock --device /dev/disk/by-id/wwn-0xdeaddeadbeef0000 sfdisk /dev/disk/by-id/wwn-0xdeaddeadbeef0000
[   21.441083] TEST-64-UDEV-STORAGE.sh[1108]: Checking that no-one is using this disk right now ... OK
[   21.442597] TEST-64-UDEV-STORAGE.sh[1108]: Disk /dev/disk/by-id/wwn-0xdeaddeadbeef0000: 16 MiB, 16777216 bytes, 32768 sectors
[   21.446893] TEST-64-UDEV-STORAGE.sh[1108]: Disk model: QEMU HARDDISK
[   21.447280] TEST-64-UDEV-STORAGE.sh[1108]: Units: sectors of 1 * 512 = 512 bytes
[   21.447552] TEST-64-UDEV-STORAGE.sh[1108]: Sector size (logical/physical): 512 bytes / 512 bytes
[   21.448169] TEST-64-UDEV-STORAGE.sh[1108]: I/O size (minimum/optimal): 512 bytes / 512 bytes
[   21.448436] TEST-64-UDEV-STORAGE.sh[1108]: >>> Script header accepted.
[   21.448847] TEST-64-UDEV-STORAGE.sh[1108]: >>> Created a new GPT disklabel (GUID: 7EB262A8-9700-45B7-B5A7-0ACAB278E869).
[   21.449186] TEST-64-UDEV-STORAGE.sh[1108]: /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part1: Created a new partition 1 of type 'Linux filesystem' and of size 5 MiB.
[   21.449632] TEST-64-UDEV-STORAGE.sh[1108]: /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part2: Created a new partition 2 of type 'Linux filesystem' and of size 5 MiB.
[   21.450134] TEST-64-UDEV-STORAGE.sh[1108]: /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part3: Done.
[   21.450483] TEST-64-UDEV-STORAGE.sh[1108]: New situation:
[   21.451038] TEST-64-UDEV-STORAGE.sh[1108]: Disklabel type: gpt
[   21.451450] TEST-64-UDEV-STORAGE.sh[1108]: Disk identifier: 7EB262A8-9700-45B7-B5A7-0ACAB278E869
[   21.451846] TEST-64-UDEV-STORAGE.sh[1108]: Device                                       Start   End Sectors Size Type
[   21.452244] TEST-64-UDEV-STORAGE.sh[1108]: /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part1  2048 12287   10240   5M Linux filesystem
[   21.452632] TEST-64-UDEV-STORAGE.sh[1108]: /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part2 12288 22527   10240   5M Linux filesystem
[   21.453024] TEST-64-UDEV-STORAGE.sh[1108]: The partition table has been altered.
[   21.453514] TEST-64-UDEV-STORAGE.sh[1108]: Calling ioctl() to re-read partition table.
[   21.453888] TEST-64-UDEV-STORAGE.sh[1108]: Syncing disks.
[   21.454339] TEST-64-UDEV-STORAGE.sh[726]: + udevadm settle
[   21.567200] TEST-64-UDEV-STORAGE.sh[726]: + udevadm lock --device /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part2 mkfs.ext4 -U deadbeef-dead-dead-beef-111111111111 -L failover_vol /dev/disk/by-id/wwn-0xdeaddeadbeef0000-part2
[   21.583437] TEST-64-UDEV-STORAGE.sh[1115]: mke2fs 1.47.1 (20-May-2024)
[   21.585429] TEST-64-UDEV-STORAGE.sh[1115]: Discarding device blocks:    0/5120         done
[   21.586605] TEST-64-UDEV-STORAGE.sh[1115]: Creating filesystem with 5120 1k blocks and 1280 inodes
[   21.588498] TEST-64-UDEV-STORAGE.sh[1115]: Allocating group tables: 0/1   done
[   21.589136] TEST-64-UDEV-STORAGE.sh[1115]: Writing inode tables: 0/1   done
[   21.592948] TEST-64-UDEV-STORAGE.sh[1115]: Creating journal (1024 blocks): done
[   21.598394] TEST-64-UDEV-STORAGE.sh[1115]: Writing superblocks and filesystem accounting information: 0/1   done
[   21.601846] TEST-64-UDEV-STORAGE.sh[726]: + modprobe -v dm_multipath
[   21.618168] TEST-64-UDEV-STORAGE.sh[1116]: insmod /lib/modules/6.11.2-arch1-1/kernel/drivers/md/dm-multipath.ko.zst
[   21.623472] TEST-64-UDEV-STORAGE.sh[726]: + systemctl start multipathd.service
[   22.917847] TEST-64-UDEV-STORAGE.sh[726]: + systemctl status multipathd.service
[   22.997215] TEST-64-UDEV-STORAGE.sh[1372]: ● multipathd.service - Device-Mapper Multipath Device Controller
[   23.030238] TEST-64-UDEV-STORAGE.sh[1372]:      Loaded: loaded (/usr/lib/systemd/system/multipathd.service; disabled; preset: disabled)
[   23.033502] TEST-64-UDEV-STORAGE.sh[1372]:      Active: active (running) since Thu 2024-10-10 22:51:07 UTC; 83ms ago
[   23.033935] TEST-64-UDEV-STORAGE.sh[1372]:  Invocation: fc00750538614228894c2b57e4597594
[   23.034360] TEST-64-UDEV-STORAGE.sh[1372]: TriggeredBy: ○ multipathd.socket
[   23.034771] TEST-64-UDEV-STORAGE.sh[1372]:    Main PID: 1123 (multipathd)
[   23.035201] TEST-64-UDEV-STORAGE.sh[1372]:      Status: "up"
[   23.035681] TEST-64-UDEV-STORAGE.sh[1372]:       Tasks: 7
[   23.036052] TEST-64-UDEV-STORAGE.sh[1372]:      Memory: 23.4M (peak: 23.5M)
[   23.036449] TEST-64-UDEV-STORAGE.sh[1372]:         CPU: 382ms
[   23.036901] TEST-64-UDEV-STORAGE.sh[1372]:      CGroup: /system.slice/multipathd.service
[   23.037328] TEST-64-UDEV-STORAGE.sh[1372]:              └─1123 /usr/bin/multipathd -d -s
[   23.038002] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbh: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.038429] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbi: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.042398] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: libmp_mapinfo: map 3deaddeadbeef0014 doesn't exist
[   23.042818] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: 3deaddeadbeef0014: addmap [0 2048 multipath 0 0 4 1 service-time 0 1 1 67:144 1 service-time 0 1 1 67:160 1 service-time 0 1 1 67:176 1 service-time 0 1 1 67:192 1]
[   23.043247] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbj: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.043768] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbk: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.044199] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbl: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.044603] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: sdbm: setting scsi timeouts is unsupported for protocol scsi:unspec
[   23.045007] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: libmp_mapinfo: map 3deaddeadbeef0015 doesn't exist
[   23.045426] TEST-64-UDEV-STORAGE.sh[1372]: Oct 10 22:51:07 H multipathd[1123]: 3deaddeadbeef0015: addmap [0 2048 multipath 0 0 4 1 service-time 0 1 1 67:208 1 service-time 0 1 1 67:224 1 service-time 0 1 1 67:240 1 service-time 0 1 1 68:0 1]
[   23.045946] TEST-64-UDEV-STORAGE.sh[726]: + multipath -ll
[   23.663065] TEST-64-UDEV-STORAGE.sh[1429]: 23.662778 | libmp_mapinfo: map 3deaddeadbeef0000-part1 has no targets
[   23.785772] integration-test-setup.sh[1479]: + set -o pipefail
[   23.786183] integration-test-setup.sh[1479]: + case "$1" in
[   23.787175] integration-test-setup.sh[1489]: ++ systemctl list-jobs
[   23.823573] integration-test-setup.sh[1479]: + [[ JOB UNIT                         TYPE  STATE
[   23.823943] integration-test-setup.sh[1479]: 727 TEST-64-UDEV-STORAGE.service start running
[   23.836220] integration-test-setup.sh[1479]: 1 jobs listed. =~ reboot.target|kexec.target|soft-reboot.target ]]
[   23.836767] integration-test-setup.sh[1479]: + rm -f /var/lib/TEST-64-UDEV-STORAGE/inprogress
[FAILED] Failed to start TEST-64-UDEV-STORAGE.service.
[   25.829654] systemd-shutdown[1]: Could not detach DM /dev/dm-9: Device or resource busy
[   25.835006] systemd-shutdown[1]: Could not detach DM /dev/dm-8: Device or resource busy
[   25.839677] systemd-shutdown[1]: Could not detach DM /dev/dm-7: Device or resource busy
[   25.847587] systemd-shutdown[1]: Could not detach DM /dev/dm-6: Device or resource busy
[   25.850149] systemd-shutdown[1]: Could not detach DM /dev/dm-5: Device or resource busy
[   25.853549] systemd-shutdown[1]: Could not detach DM /dev/dm-4: Device or resource busy
[   25.854911] systemd-shutdown[1]: Could not detach DM /dev/dm-3: Device or resource busy
[   25.857695] systemd-shutdown[1]: Could not detach DM /dev/dm-2: Device or resource busy
[   25.862971] systemd-shutdown[1]: Could not detach DM /dev/dm-17: Device or resource busy
[   25.867429] systemd-shutdown[1]: Could not detach DM /dev/dm-16: Device or resource busy
[   25.876760] systemd-shutdown[1]: Could not detach DM /dev/dm-15: Device or resource busy
[   25.881057] systemd-shutdown[1]: Could not detach DM /dev/dm-14: Device or resource busy
[   25.890508] systemd-shutdown[1]: Could not detach DM /dev/dm-13: Device or resource busy
[   25.894483] systemd-shutdown[1]: Could not detach DM /dev/dm-12: Device or resource busy
[   25.898254] systemd-shutdown[1]: Could not detach DM /dev/dm-11: Device or resource busy
[   25.900154] systemd-shutdown[1]: Could not detach DM /dev/dm-10: Device or resource busy
[   25.905220] systemd-shutdown[1]: Could not detach DM /dev/dm-1: Device or resource busy
[   25.907517] systemd-shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[   25.920448] systemd-shutdown[1]: Unable to finalize remaining DM devices, ignoring.
[   25.962978] reboot: Power down
Finishing after writing 78469 entries
‣ Setting UseSubvolumes should be configured in [Build], not [Output].
‣ Setting WorkspaceDirectory should be configured in [Build], not [Output].
‣ Setting PackageCacheDirectory should be configured in [Build], not [Output].
‣ Setting Environment should be configured in [Build], not [Content].
Test failed, relevant logs can be viewed with: 

gh run download 11282993419 --name ci-mkosi-11282993419-1-arch-rolling-failed-test-journals -D ci/ci-mkosi-11282993419-1-arch-rolling-failed-test-journals && journalctl --file ci/ci-mkosi-11282993419-1-arch-rolling-failed-test-journals/test/journal/TEST-64-UDEV-STORAGE-multipath_basic_failover.journal --no-hostname -o short-monotonic -u TEST-64-UDEV-STORAGE.service -p info
―――――――��――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```
</details>

File: src/udev/udev-node.c [hunk 2]
```diff
@@ -360,45 +360,59 @@ static int stack_directory_get_name(const char *slink, char **ret) {
...
 
-        r = stack_directory_get_name(slink, &dirname);
+        r = stack_directory_get_name(slink, &name);
         if (r < 0)
                 return log_device_debug_errno(dev, r, "Failed to build stack directory name for '%s': %m", slink);
 
-        r = mkdir_parents(dirname, 0755);
-        if (r < 0)
-                return log_device_debug_errno(dev, r, "Failed to create stack directory '%s': %m", dirname);
+        FOREACH_STRING(s, "/run/udev/links/", "/run/udev/.links.lock/") {
```
Comment (poettering) [inline src/udev/udev-node.c]: why hide the lock dir? this is our internal stuff anyway, no need to hide this. I'd just call this `/run/udev/links.lock/` hence.

File: src/udev/udev-node.c [hunk 3]
```diff
@@ -360,45 +360,59 @@ static int stack_directory_get_name(const char *slink, char **ret) {
...
-        if (dirfd < 0)
-                return log_device_debug_errno(dev, dirfd, "Failed to open stack directory '%s': %m", dirname);
+        /* 1. Take a lock for the stack directory. */
+        lockname = path_join("/run/udev/links.lock/", name);
+        if (!lockname)
+                return -ENOMEM;
+
+        r = make_lock_file(lockname, LOCK_EX, &lockfile);
+        if (r < 0)
+                return log_device_debug_errno(dev, errno, "Failed to create lock file '%s': %m", lockname);
```
Comment (YHNdnzj) [inline src/udev/udev-node.c]: errno -> r?

**Pull request merged**